include/seqan3/alphabet/quality/quality_base.hpp

```diff
@@ -0,0 +1,195 @@
+    /*!\name Member types
+     * \{
+     */
+    //!\brief The integer representation of a quality score assignable with =operator.
+    using phred_type = int8_t;
+    //!\}
+
+private:
+    //!\brief The base type.
+    using base_t = alphabet_base<derived_type, size, char_t>;
+
+    /*!\name Constructors, destructor and assignment
+     * \{
+     */
+    constexpr quality_base() : base_t{} {}
+    constexpr quality_base(quality_base const &) = default;
+    constexpr quality_base(quality_base &&) = default;
+    constexpr quality_base & operator=(quality_base const &) = default;
+    constexpr quality_base & operator=(quality_base &&) = default;
+
+    constexpr quality_base(phred_type const p) noexcept
+    {
+        static_cast<derived_type *>(this)->assign_phred(p);
+    }
+    //!\}
+
+    //!\brief Befriend the derived_type so it can instantiate.
+    friend derived_type;
+
+public:
+    // Import from base type:
+    using base_t::value_size;
+    using base_t::to_rank;
+    using base_t::assign_rank;
+    using typename base_t::char_type;
+    using typename base_t::rank_type;
+
+    /*!\name Read functions
+     * \{
+     */
+    constexpr char_type to_char() const noexcept
+    {
+        return static_cast<char_type>(to_rank()) + derived_type::offset_char;
+    }
+
+    constexpr phred_type to_phred() const noexcept
+    {
+        return static_cast<phred_type>(to_rank()) + derived_type::offset_phred;
+    }
+    //!\}
+
+    /*!\name Write functions
+     * \{
+     */
+
+    /*!\brief Assign from the numeric phred value.
+     *
+     * \details
+     *
+     * Satisfies the seqan3::quality_concept::assign_phred() requirement via the seqan3::assign_rank() wrapper.
+     *
+     * \par Complexity
+     *
+     * Constant.
+     */
+    constexpr derived_type & assign_phred(phred_type const p) noexcept
+    {
+        return assign_rank(phred_to_rank[static_cast<rank_type>(p)]);
+    }
+    //!\}
+
+    //!\brief Explicit conversion to any other quality alphabet (via char representation).
+    //!\tparam other_qual_type The type to convert to; must satisfy seqan3::quality_concept.
+    //
+    template <typename other_qual_type>
+    //!\cond
+        requires !std::Same<other_qual_type, derived_type> && quality_concept<other_qual_type>
+    //!\endcond
+    explicit constexpr operator other_qual_type() const noexcept
+    {
+        return other_qual_type{to_phred()};
+    }
+    //!\}
+
+protected:
+    //!\brief Phred to rank conversion table.
+    static std::array<rank_type, 256> constexpr phred_to_rank
+    {
+        [] () constexpr
+        {
+            std::array<rank_type, 256> ret{};
+
+            phred_type min = derived_type::offset_phred;
+            phred_type sup = min + value_size;
+
+            // map too small to smallest possible
+            for (int64_t i = std::numeric_limits<phred_type>::lowest(); i < min ; ++i)
+                ret[static_cast<rank_type>(i)] = - derived_type::offset_phred + min;
```

Comment:
> why static cast into rank_type? and not size_t?

Because underflow will result in values out of the range

> what's up with the order?

has changed already

Reply:
> > why static cast into rank_type? and not size_t?
> 
> Because underflow will result in values out of the range

I know, but is rank_type always unsigned? Has it the same size as phred_type? As someone who only sees this code it is not apparent which assumptions hold here. And therefore the confused question from me. It would be more clear if you use `std::make_unsigned_t<phred_type>`.